### PR TITLE
Memory units

### LINKED ISSUE #1435: Failed to set memory requirements in kafka spec
I changed the kafka spec to include the following:

```
spec:
  kafka:
    resources:
      requests:
        cpu: 500m
        memory: 3500Mi
```

And I got the following error:

```2019-03-12 14:20:34 ERROR AbstractAssemblyOperator:176 - Reconciliation #85(timer) Kafka(strimzi/strimzi-cluster): createOrUpdate failed
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://k8ssetupdev-c13169a8.hcp.westeurope.azmk8s.io/apis/apps/v1/namespaces/strimzi/statefulsets. Message: StatefulSet in version "v1" cannot be handled as a StatefulSet: v1.StatefulSet.Spec: v1.StatefulSetSpec.Template: v1.PodTemplateSpec.Spec: v1.PodSpec.Containers: []v1.Container: v1.Container.VolumeMounts: []v1.VolumeMount: Resources: v1.ResourceRequirements.Requests: unmarshalerDecoder: unable to parse quantity's suffix, error found in #10 byte of ...|"3670016K"}},"volume|..., bigger context ...|s":{},"requests":{"cpu":"500m","memory":"3670016K"}},"volumeMounts":[{"mountPath":"/var/lib/kafka/da|.... Received status: Status(apiVersion=v1, code=400, details=null, kind=Status, message=StatefulSet in version "v1" cannot be handled as a StatefulSet: v1.StatefulSet.Spec: v1.StatefulSetSpec.Template: v1.PodTemplateSpec.Spec: v1.PodSpec.Containers: []v1.Container: v1.Container.VolumeMounts: []v1.VolumeMount: Resources: v1.ResourceRequirements.Requests: unmarshalerDecoder: unable to parse quantity's suffix, error found in #10 byte of ...|"3670016K"}},"volume|..., bigger context ...|s":{},"requests":{"cpu":"500m","memory":"3670016K"}},"volumeMounts":[{"mountPath":"/var/lib/kafka/da|..., metadata=ListMeta(_continue=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=BadRequest, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:478) ~[cluster-operator-0.11.1.jar:0.11.1]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:417) ~[cluster-operator-0.11.1.jar:0.11.1]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:381) ~[cluster-operator-0.11.1.jar:0.11.1]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:344) ~[cluster-operator-0.11.1.jar:0.11.1]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleCreate(OperationSupport.java:227) ~[cluster-operator-0.11.1.jar:0.11.1]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleCreate(BaseOperation.java:812) ~[cluster-operator-0.11.1.jar:0.11.1]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:382) ~[cluster-operator-0.11.1.jar:0.11.1]
	at io.strimzi.operator.common.operator.resource.AbstractResourceOperator.internalCreate(AbstractResourceOperator.java:165) ~[cluster-operator-0.11.1.jar:0.11.1]
	at io.strimzi.operator.cluster.operator.resource.StatefulSetOperator.internalCreate(StatefulSetOperator.java:271) ~[cluster-operator-0.11.1.jar:0.11.1]
	at io.strimzi.operator.cluster.operator.resource.StatefulSetOperator.internalCreate(StatefulSetOperator.java:39) ~[cluster-operator-0.11.1.jar:0.11.1]
	at io.strimzi.operator.common.operator.resource.AbstractResourceOperator.lambda$reconcile$0(AbstractResourceOperator.java:91) ~[cluster-operator-0.11.1.jar:0.11.1]
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:273) ~[cluster-operator-0.11.1.jar:0.11.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_191]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_191]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [cluster-operator-0.11.1.jar:0.11.1]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_191]```

### PR DESCRIPTION
**Pull request by tombentley**
### Type of change

- Bugfix
- Refactoring

### Description

Fix #1435. For historical reasons the CO was parsing and reformatting memory requests & limits and using the `K` suffix. I think the bug is due to poor parsing in k8s (it accepts lower case `k`, even though upper case `K` is documented as acceptable). But on investigating I realised that the CO no longer makes use of the actual value (it did at one point...), so there's no point in normalising at all. But then we might as well use the standard k8s schema (i.e. Fabric8 `ResourceRequirements`). Since resource requirements are not _necessarily_ limited to `memory` and `cpu`, this means users are specify other requirements.

Note this PR also adds support for suffixed memory with decimal quantities (`1.1Gi`) and even the crazy-looking `1e3Mi`, but since the k8s documentation is a little vague about what is acceptable I erred on the side of laxness.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: api/src/main/java/io/strimzi/api/kafka/model/EntityTopicOperatorSpec.java [hunk 1]
```diff
@@ -102,12 +103,12 @@ public void setTopicMetadataMaxAttempts(int topicMetadataMaxAttempts) {
     }
 
     @Description("Resource constraints (limits and requests).")
-    public Resources getResources() {
+    public ResourceRequirements getResources() {
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/EntityTopicOperatorSpec.java]: Should we use the `@KubeLink` annotation to link it to the Kubernetes API reference?

**Comment by tombentley** [comment]
Something in k8s seems to update the STS. So, for example if you set memory 123000 on the STS initially it gets updated to 123k by k8s. This PR changes nothing about that.

**Comment by scholzj** [comment]
Yeah, but the important part is using the same units to make sure it pops up in the diffs only when it is really different. Because otherwise we:
* Might trigger a rolling update every reconciliation just because the units are different on K8S side
* We don't update the STS when the user changes the limits

I think we had that solved before, right?

**Comment by tombentley** [comment]
The only logic in `StatefulSetDiff` is to simply ignore the resources, so nothing changes there, so nothing should break either.

**Comment by scholzj** [comment]
And how do we trigger rolling update when the resources change?

**Comment by tombentley** [comment]
IDK, but are you sure this actually works ATM?

**Comment by scholzj** [comment]
Well, that is what I was wondering about ... but I was under the impression that it did worked ... I will try it.

**Comment by scholzj** [comment]
Ok, so current master does rolling update when resources change.

**Comment by tombentley** [comment]
And it works with these changes too. So if you're happy, I'll merge this.

**Comment by scholzj** [comment]
@tombentley Yeah, if it works, merge it.

**Pull request merged**